File: java/api/model/src/main/java/com/vmware/photon/controller/api/model/VmNetworks.java
/*
 * Copyright 2015 VMware, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy of
 * the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, without warranties or
 * conditions of any kind, EITHER EXPRESS OR IMPLIED.  See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.vmware.photon.controller.api.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.wordnik.swagger.annotations.ApiModel;
import com.wordnik.swagger.annotations.ApiModelProperty;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

/**
 * Networks of VM.
 */
@ApiModel(value = "This class represents networks of a VM.")
@JsonIgnoreProperties(ignoreUnknown = true)
public class VmNetworks {

  @JsonProperty
  @ApiModelProperty(value = "Network connections of VM")
  private Set<NetworkConnection> networkConnections = new HashSet<>();

  public Set<NetworkConnection> getNetworkConnections() {
    return networkConnections;
  }

  public void setNetworkConnections(Set<NetworkConnection> networkConnections) {
    this.networkConnections = networkConnections;
  }

  /**
   * Add a networkConnection to a VM's networks.
   *
   * @param networkConnection the network connection
   * @return <tt>true</tt> if this network did not already contain the specified
   * networkConnection. If this network already contains the networkConnection,
   * the call leaves it unchanged and returns <tt>false</tt>
   */
  public boolean addNetworkConnection(NetworkConnection networkConnection) {
    return networkConnections.add(networkConnection);
  }

  /**
   * Remove a networkConnection from a VM's networks.
   *
   * @param networkConnection the network connection
   * @return <tt>true</tt> if this network contained the specified network,
   * otherwise returns <tt>false</tt>
   */
  public boolean removeNetworkConnection(NetworkConnection networkConnection) {
    return networkConnections.remove(networkConnection);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    VmNetworks other = (VmNetworks) o;

    return Objects.equals(networkConnections, other.networkConnections);
  }

  @Override
  public int hashCode() {
    return Objects.hash(networkConnections);
  }
}

File: tests/common/lib/common/resources/vm_networks.rb
# Copyright 2015 VMware, Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License. You may obtain a copy of
# the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, without warranties or
# conditions of any kind, EITHER EXPRESS OR IMPLIED. See the License for the
# specific language governing permissions and limitations under the License.

module EsxCloud
  class VmNetworks

    attr_reader :network_connections

    # @param [Task] task
    # @return [VmNetworks]
    def self.create_from_task(task)
      resource_properties = task.resource_properties
      networks = resource_properties["networkConnections"]
      unless networks.is_a?(Array)
        fail UnexpectedFormat, "Invalid networks: #{networks.inspect}"
      end
      networks.each do |network|
        unless network.is_a?(Hash)
          fail UnexpectedFormat, "Invalid network format: #{network.inspect}"
        end
      end

      return VmNetworks.create_from_hash_array(networks)
    end

    # @param [Array] networks
    # @return [VmNetworks]
    def self.create_from_hash_array(networks)
      unless networks.is_a?(Array)
        fail UnexpectedFormat, "Invalid hash_array: #{networks.inspect}"
      end
      network_connections = networks.map do |network|
        unless network.is_a?(Hash)
          fail UnexpectedFormat, "Invalid network: #{network.inspect}"
        end
        NetworkConnection.new(network["network"],
                             network["macAddress"],
                             network["ipAddress"],
                             network["netmask"],
                             network["isConnected"])
      end

      new(network_connections)
    end

    # @param [Hash] network_connections
    def initialize(network_connections)
      @network_connections = network_connections
    end
  end
end
